openedx/core/djangoapps/course_groups/models.py

```diff
@@ -92,52 +87,47 @@ def clean(self):
         if self.course_user_group.course_id != self.course_id:
             raise ValidationError("Non-matching course_ids provided")
 
-    def save(self, *args, **kwargs):
-        self.full_clean(validate_unique=False)
-
-        log.info("Saving CohortMembership for user '%s' in '%s'", self.user.id, self.course_id)
-
-        # Avoid infinite recursion if creating from get_or_create() call below.
-        # This block also allows middleware to use CohortMembership.get_or_create without worrying about outer_atomic
-        if 'force_insert' in kwargs and kwargs['force_insert'] is True:
-            with transaction.atomic():
-                self.course_user_group.users.add(self.user)
-                super(CohortMembership, self).save(*args, **kwargs)
-            return
-
-        # This block will transactionally commit updates to CohortMembership and underlying course_user_groups.
-        # Note the use of outer_atomic, which guarantees that operations are committed to the database on block exit.
-        # If called from a view method, that method must be marked with @transaction.non_atomic_requests.
-        with outer_atomic(read_committed=True):
-
-            saved_membership, created = CohortMembership.objects.select_for_update().get_or_create(
-                user__id=self.user.id,
-                course_id=self.course_id,
+    @classmethod
+    def assign(cls, cohort, user):
+        """
+        Assign user to cohort, switching them to this cohort if they had previously been assigned to another
+        cohort
+        Returns CohortMembership, previous_cohort (if any)
+        """
+        with transaction.atomic():
+            membership, created = cls.objects.select_for_update().get_or_create(
+                user__id=user.id,
```

Comment:
nit: Since you're passing in an instance of `User`, you should be able to write this as `user=user`.